libs/components/packages/src/schematics/migrations/update-7/ag-grid-28.schematic.ts

```diff
@@ -0,0 +1,195 @@
+import { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
+import { NodePackageInstallTask } from '@angular-devkit/schematics/tasks';
+import {
+  NodeDependencyType,
+  addPackageJsonDependency,
+  getPackageJsonDependency,
+  removePackageJsonDependency,
+} from '@schematics/angular/utility/dependencies';
+
+import ignore from 'ignore';
+import { relative } from 'path';
+import semver from 'semver';
+
+export const UPDATE_TO_VERSION = '28.2.0';
+
+const UPDATE_LT_VERSION = '28.0.0';
+const ANY_MODULE = '@ag-grid-community/';
+const ENT_MODULE = '@ag-grid-enterprise/';
+const AG_GRID = 'ag-grid-community';
+const AG_GRID_ENT = 'ag-grid-enterprise';
+const withComponents = 'AgGridModule.withComponents';
+const warned: string[] = [];
+
+function warnOnce(message: string) {
+  if (!warned.includes(message)) {
+    warned.push(message);
+    console.warn(message);
+  }
+}
+
+/**
+ * Upgrade to AG Grid 28 and address three breaking changes:
+ *
+ * - No longer support `@ag-grid-community/all-modules` -- just use the package
```

Comment:
👇 For readability, it would be beneficial to split these three fixes into separate functions that return a `Rule`. (You don't need to use these names; they're just for demonstration purposes.)
```
async function fixAgGridImports(): Rule {
  return (tree) => {...}
}
// etc...
```
Then, in your default function you can return a `chain` of the rules:
```
export default function (): Rule {
  return (tree, context) => {
    return chain([
      fixAgGridImports(),
      removeWithComponentsModuleImport(),
      renameColumnApi()
    ]);
  };
}
```